python/cudf/cudf/core/udf/lowering.py

```diff
@@ -0,0 +1,131 @@
+@cuda_lower(operator.add, MaskedType, MaskedType)
+def masked_scalar_add_impl(context, builder, sig, args):
+    '''
+    Implement `MaskedType` + `MaskedType`
+    '''
+
+    masked_type_1, masked_type_2 = sig.args # MaskedType(...), MaskedType(...)
+    masked_return_type = sig.return_type # MaskedType(...)
+
+    # Let there be two actual LLVM structs backing the two inputs
+    # https://mapping-high-level-constructs-to-llvm-ir.readthedocs.io/en/latest/basic-constructs/structures.html
+    m1 = cgutils.create_struct_proxy(masked_type_1)(
+        context, builder, value=args[0]
+    )
+    m2 = cgutils.create_struct_proxy(masked_type_2)(
+        context, builder, value=args[1]
+    )
+
+    # we will return an output struct
+    result = cgutils.create_struct_proxy(masked_return_type)(context, builder)
+
+    # compute output validity
+    valid = builder.and_(m1.valid, m2.valid)
+    result.valid = valid
+    with builder.if_then(valid):
+        # Let numba handle generating the extra LLVM needed to perform
+        # operations on mixed types, by compiling the final core op between
+        # the two primitive values as a separate function and calling it
+        result.value = context.compile_internal(
+            builder, 
+            lambda x, y: x + y, 
+            nb_signature(
+                masked_return_type.value_type, 
+                masked_type_1.value_type, 
+                masked_type_2.value_type
+            ),
+            (m1.value, m2.value)
+        )
+    return result._getvalue()
+
+@cuda_lower(operator.add, MaskedType, NAType)
+def masked_scalar_add_na_impl(context, builder, sig, args):
+    '''
+    Implement `MaskedType` + `NAType`
+    The answer to this is known up front so no actual addition
+    needs to take place
+    '''
+
+    return_type = sig.return_type # MaskedType(...)
+    result = cgutils.create_struct_proxy(MaskedType(return_type.value_type))(
+        context, builder
+    )
+
+    # Invalidate the struct and leave `value` uninitialized
+    result.valid = context.get_constant(types.boolean, 0)
+    return result._getvalue()
+
+@cuda_lower(operator.add, MaskedType, types.Integer)
+def masked_scalar_add_constant_impl(context, builder, sig, input_values):
+    '''
+    Implement `MaskedType` + constant
+    '''
+    masked_type, const_type = sig.args
+
+    indata = cgutils.create_struct_proxy(MaskedType(masked_type.value_type))(
+        context, builder, value=input_values[0]
+    )
+    result = cgutils.create_struct_proxy(MaskedType(masked_type.value_type))(
+        context, builder
+    )
+    result.valid = context.get_constant(types.boolean, 0)
+    with builder.if_then(indata.valid):
+        result.value = builder.add(indata.value, input_values[1])
+        result.valid = context.get_constant(types.boolean, 1)
+
+    return result._getvalue()
+
+@cuda_lower(operator.is_, MaskedType, NAType)
+def masked_scalar_is_null_impl(context, builder, sig, args):
```

Comment:
@gmarkall any idea why this seems to always return `1`? Popping up a few levels to `lower_binop` I am seeing this:

```
# lowering.py(109)
(Pdb) impl(self.builder, (lhs, rhs))
<ir.Constant type='i1' value=1>
```

Reply:
The last thing the code did was `result = context.get_constant(types.boolean, 1)` - both sides of the branch "execute" in Python to build the IR, so here you create a constant `1` and return it. To implement this correctly you need to allocate space on the stack for the result, store the relevant value into it in each branch, and then return the loaded value.

e.g. something like

```
result = cgutils.alloca_once(builder)

with builder.if_else(indata.valid) as (then, otherwise):
    with then:
        builder.store(context.get_constant(types.boolean, 0), result)
    with otherwise:
        builder.store(context.get_constant(types.boolean, 1), result)

return builder.load(result)
```

(Apologies for missing this when I read the code before).